### PR TITLE
allow user to add multiple unique items from clipboard

### PR DESCRIPTION
**Pull request by maheedhar1305**

### CODE REVIEW

**Comment by maheedhar1305** [comment]
@blackjackkent I have made the changes in the new version of the code and added a couple of unit tests. Please review when you get some time. Thanks again for your help.

**Comment by blackjackkent** [comment]
This is looking great! A couple small things -

1) Please run `npm run validate`; there's a few small linting issues to address. (The new dev setup should have installed some hooks to check for this on commit/push with husky but I think since you already had the repo cloned, it didn't run them or something.)
2) One edge case I see here is that since we moved to `event.key`, text that (for instance) contains the term 'Enter' will get split even though that's likely not the user's intention. So you might want to add some logic to strip out non-character key labels like that. (Info at https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key and https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values)

Other than that, this looks great as far as I can tell. Thank you for adding the tests as well!

**Comment by maheedhar1305** [comment]
- I fixed the es lint issues.
- Good catch on the edge case. How would you strip out non-character key labels in this case? Should i just remove non character key labels from the delimiters list I use for splitting the text into terms?

**Comment by blackjackkent** [comment]
> * Good catch on the edge case. How would you strip out non-character key labels in this case? Should i just remove non character key labels from the delimiters list I use for splitting the text into terms?

Yeah - probably a couple ways to do this but my strategy would probably be to have a constant list of delimiter labels to skip, and when you're looping through `submitKeys`, just skip any that are in that list. :)

**Comment by maheedhar1305** [comment]
I went through the list of keys list here : https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values and felt like Enter and Tab where the only keys folks might try to use as a submit key so I only used that instead of making a huge list of all the keys. Please let me know if that is sufficient. Thanks

File: src/components/MultipleValueTextInput.tsx [hunk 1]
```diff
@@ -64,6 +66,24 @@ function MultipleValueTextInput({
 		setValue('');
 		onItemAdded(value, newValues);
 	};
+	const handleItemsAdd = (addedValues: string[]) => {
+		let uniqueValues: string[] = []
```
Comment (blackjackkent) [inline src/components/MultipleValueTextInput.tsx]: A Javascript tidbit for the future - since `addedValues` will simply be an array of strings, we can use the following to remove duplicates from it:

```
let uniqueValues = Array.from(new Set(addedValues));
```

which would reduce lines 70-75 down to one line. 😄 

We also could (and maybe should?) do the same thing on line 77 to make sure there's no duplicates between the pasted values and the existing array of values.

```
const newValues = Array.from(new Set([...values, ...uniqueValues]));
```

(That last bit is the "spread" operator syntax; happy to provide more resources on that too.)

File: src/components/MultipleValueTextInput.tsx [hunk 2]
```diff
@@ -84,6 +104,33 @@ function MultipleValueTextInput({
...
+		let result: string = str
+		for (let i = 1; i < delimiters.length; i+=1) {
+			result = result.split(delimiters[i]).join(tempChar) // Handle scenarios where pasted text has more than one submitKeys in it
+		}
+		return result.split(tempChar)
+	}
+
+	const handlePaste = (e: React.ClipboardEvent<HTMLInputElement>) => {
+		const pastedText = e.clipboardData.getData('text/plain')
+		let areSubmitKeysPresent = false
```
Comment (blackjackkent) [inline src/components/MultipleValueTextInput.tsx]: Another fun bit of Javascript array syntax to play with - lines 119-124 could be replaced with:

```
const areSubmitKeysPresent = delimiters.some((d) => pastedText.includes(d));
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

**Review by blackjackkent — DISMISSED** [review]
This looks good and tests look great. :) Added a few comments on some Javascript syntax tweaks, but those can be addressed in the future. I will merge this later tonight unless you want to make those changes now; I know you have another project waiting on this. 😄

**Comment by blackjackkent** [comment]
> I went through the list of keys list here : https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values and felt like Enter and Tab where the only keys folks might try to use as a submit key so I only used that instead of making a huge list of all the keys. Please let me know if that is sufficient.

This makes sense and works for me. :D

File: src/components/MultipleValueTextInput.tsx (hunk 1 above)
Comment (maheedhar1305) [reply src/components/MultipleValueTextInput.tsx]: oh let me tc of this real quick

Comment (maheedhar1305) [reply src/components/MultipleValueTextInput.tsx]: jeez, I was so caught up in trying to get the biz logic to work within react/js, that I forgot i was basically implementing a set in lines 70-75. so dumb!!! Thanks hehe

Comment (maheedhar1305) [reply src/components/MultipleValueTextInput.tsx]: you actually caught a bug there in line 77 too where we would have had duplicates if values and uniqueValues had the same item in them. good eye

Comment (blackjackkent) [reply src/components/MultipleValueTextInput.tsx]: I wouldn't call it dumb. :D Your approach was valid (and the syntax I described is comparatively new in JS). But I do like the conciseness of the more modern JS array functions (as long as they don't interfere with readability).

Comment (maheedhar1305) [reply src/components/MultipleValueTextInput.tsx]: ya kinda like what java does with collections/streams. i dig it

Comment (maheedhar1305) [reply src/components/MultipleValueTextInput.tsx]: ok I took your advice and had to slightly tweak it 
I replaced lines 70-75 with
 `let uniqueValues = Array.from(new Set(addedValues.filter(elm => elm && !values.includes(elm))));`
 
 because in addition to making sure the elements are unique, I also need to remove empty/null and also remove items that are already in the values array. I need to do this here because I then use uniqueValues to call onItemAdded and so I cannot have existing values in this array. Let me know if there is a better way to do this

Comment (blackjackkent) [reply src/components/MultipleValueTextInput.tsx]: Nope, that looks great!

**Comment by blackjackkent** [comment]
Thanks for taking a look at these changes. :) I think we're all good to go here; gonna run CI stuff and get this merged.
